### ISSUE TITLE
Install build dependencies error

### ISSUE THREAD

**Issue by lexaSvarshik** [openzim/openedx#169]
I got two systems (Win10 and Linux, on different computers), and I'm getting the same error during the installation. Here's log:

`  Installing build dependencies ... error
  error: subprocess-exited-with-error

  × pip subprocess to install build dependencies did not run successfully.
  │ exit code: 1
  ╰─> [83 lines of output]
      Collecting setuptools>=35.0.2
        Using cached setuptools-65.6.3-py3-none-any.whl (1.2 MB)
      Collecting wheel>=0.29.0
        Using cached wheel-0.38.4-py3-none-any.whl (36 kB)
      Collecting twine
        Using cached twine-4.0.2-py3-none-any.whl (36 kB)
      Collecting cython==0.29.6
        Using cached Cython-0.29.6.tar.gz (2.0 MB)
        Preparing metadata (setup.py): started
        Preparing metadata (setup.py): finished with status 'error'
        error: subprocess-exited-with-error
     
        python setup.py egg_info did not run successfully.
        exit code: 1
     
        [56 lines of output]
        Unable to find pgen, not compiling formal grammar.
        running egg_info
        creating C:\Users\wired_god\AppData\Local\Temp\pip-pip-egg-info-09q3qd_w\Cython.egg-info
        writing C:\Users\wired_god\AppData\Local\Temp\pip-pip-egg-info-09q3qd_w\Cython.egg-info\PKG-INFO
        writing dependency_links to C:\Users\wired_god\AppData\Local\Temp\pip-pip-egg-info-09q3qd_w\Cython.egg-info\dependency_links.txt
        writing entry points to C:\Users\wired_god\AppData\Local\Temp\pip-pip-egg-info-09q3qd_w\Cython.egg-info\entry_points.txt
        writing top-level names to C:\Users\wired_god\AppData\Local\Temp\pip-pip-egg-info-09q3qd_w\Cython.egg-info\top_level.txt
        writing manifest file 'C:\Users\wired_god\AppData\Local\Temp\pip-pip-egg-info-09q3qd_w\Cython.egg-info\SOURCES.txt'
        Traceback (most recent call last):
          File "<string>", line 2, in <module>
          File "<pip-setuptools-caller>", line 34, in <module>
          File "C:\Users\wired_god\AppData\Local\Temp\pip-install-dbwt20d9\cython_a0980e6b7bd24a518c503c8115eea73f\setup.py", line 228, in <module>
            setup(
          File "C:\Users\wired_god\AppData\Local\Programs\Python\Python311\Lib\site-packages\setuptools\__init__.py", line 87, in setup
            return distutils.core.setup(**attrs)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          File "C:\Users\wired_god\AppData\Local\Programs\Python\Python311\Lib\site-packages\setuptools\_distutils\core.py", line 185, in setup
            return run_commands(dist)
                   ^^^^^^^^^^^^^^^^^^
          File "C:\Users\wired_god\AppData\Local\Programs\Python\Python311\Lib\site-packages\setuptools\_distutils\core.py", line 201, in run_commands
            dist.run_commands()
          File "C:\Users\wired_god\AppData\Local\Programs\Python\Python311\Lib\site-packages\setuptools\_distutils\dist.py", line 969, in run_commands
            self.run_command(cmd)
          File "C:\Users\wired_god\AppData\Local\Programs\Python\Python311\Lib\site-packages\setuptools\dist.py", line 1208, in run_command
            super().run_command(command)
          File "C:\Users\wired_god\AppData\Local\Programs\Python\Python311\Lib\site-packages\setuptools\_distutils\dist.py", line 988, in run_command
            cmd_obj.run()
          File "C:\Users\wired_god\AppData\Local\Programs\Python\Python311\Lib\site-packages\setuptools\command\egg_info.py", line 308, in run
            self.find_sources()
          File "C:\Users\wired_god\AppData\Local\Programs\Python\Python311\Lib\site-packages\setuptools\command\egg_info.py", line 316, in find_sources
            mm.run()
          File "C:\Users\wired_god\AppData\Local\Programs\Python\Python311\Lib\site-packages\setuptools\command\egg_info.py", line 560, in run
            self.add_defaults()
          File "C:\Users\wired_god\AppData\Local\Programs\Python\Python311\Lib\site-packages\setuptools\command\egg_info.py", line 597, in add_defaults
            sdist.add_defaults(self)
          File "C:\Users\wired_god\AppData\Local\Programs\Python\Python311\Lib\site-packages\setuptools\command\sdist.py", line 106, in add_defaults
            super().add_defaults()
          File "C:\Users\wired_god\AppData\Local\Programs\Python\Python311\Lib\site-packages\setuptools\_distutils\command\sdist.py", line 252, in add_defaults
            self._add_defaults_ext()
          File "C:\Users\wired_god\AppData\Local\Programs\Python\Python311\Lib\site-packages\setuptools\_distutils\command\sdist.py", line 336, in _add_defaults_ext
            build_ext = self.get_finalized_command('build_ext')
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          File "C:\Users\wired_god\AppData\Local\Programs\Python\Python311\Lib\site-packages\setuptools\_distutils\cmd.py", line 305, in get_finalized_command
          File "C:\Users\wired_god\AppData\Local\Programs\Python\Python311\Lib\site-packages\setuptools\_distutils\cmd.py", line 111, in ensure_finalized
            self.finalize_options()
          File "C:\Users\wired_god\AppData\Local\Temp\pip-install-dbwt20d9\cython_a0980e6b7bd24a518c503c8115eea73f\Cython\Distutils\build_ext.py", line 20, in finalize_options
            self.distribution.ext_modules[:] = cythonize(
                                               ^^^^^^^^^^
          File "C:\Users\wired_god\AppData\Local\Temp\pip-install-dbwt20d9\cython_a0980e6b7bd24a518c503c8115eea73f\Cython\Build\Dependencies.py", line 960, in cythonize
            module_list, module_metadata = create_extension_list(
                                           ^^^^^^^^^^^^^^^^^^^^^^
          File "C:\Users\wired_god\AppData\Local\Temp\pip-install-dbwt20d9\cython_a0980e6b7bd24a518c503c8115eea73f\Cython\Build\Dependencies.py", line 754, in create_extension_list
            elif isinstance(patterns, basestring) or not isinstance(patterns, collections.Iterable):
                                                                              ^^^^^^^^^^^^^^^^^^^^
        AttributeError: module 'collections' has no attribute 'Iterable'
        [end of output]
     
        note: This error originates from a subprocess, and is likely not a problem with pip.
      error: metadata-generation-failed
     
      Encountered error while generating package metadata.
     
      See above for output.
     
      note: This is an issue with the package mentioned above, not pip.
      hint: See above for details.
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× pip subprocess to install build dependencies did not run successfully.
│ exit code: 1
╰─> See above for output.
`

I already got Cython 0.29.32 and MSVC compiler, so, I guess, it may be a bug.

**Comment by kelson42** [comment]
This scraper does not support Windows yet, do you have your log on Linux?

**Comment by rgaudin** [comment]
As @kelson42 pointed out, Windows is not supported.

This particular error is because:
- you are using Python3.11 
- zimscraperlib referenced in requirements is relatively old
- it references a relatively old version of libzim
- there is no wheel for that version and python3.11 (and we don't build windows ones anyway)
- so it tries to compile it using Cython
- the Cython version referenced from that libzim version's source isn't compatible with python3.11

In short, use python3.8 on linux.

**Comment by kelson42** [comment]
@rgaudin There is no way to check the python version early and gove a proper error if not matching requirement?

**Comment by rgaudin** [comment]
Well yes and no.

It is possible to mark a maximum Python version for a package and the package manager would take care of the rest (here the error is when pip is installing Cython). We do it for pylibzim now but we haven't always been (and the version referenced here was before that change).

We don't do it on scrapers because our code is non-exotic pure-python code and that's mostly forward compatible. It's not the scraper code that's not compatible with py311 here ; it's a dependency (of a dependency of a dependency) so that's its responsibility.

Also, and that's important ; we're Docker-first. Python usage is mostly reserved for development and a python developer would have noticed that the README lists external dependencies for Linux and macOS but not Windows. A python developer would also have checked the Dockerfile and noted that it's using py38 so that would be the best version to test against. We could definitely do better in the README regarding all this.
